core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala

```diff
@@ -74,20 +74,26 @@ trait TestPrematureExit {
     @volatile var exitedCleanly = false
     mainObject.exitFn = (_) => exitedCleanly = true
 
+    var message: String = null
     val thread = new Thread {
       override def run() = try {
         mainObject.main(input)
       } catch {
         // If exceptions occur after the "exit" has happened, fine to ignore them.
         // These represent code paths not reachable during normal execution.
-        case e: Exception => if (!exitedCleanly) throw e
+        case e: Exception =>
+          message = e.getMessage
+          if (!(exitedCleanly || message.contains(searchString))) {
```

Comment:
If it didn't exit cleanly, can it be possible that the exception is the correct expected one, and that its message contains the search string? I'm probably missing the reason why this has to be checked here.

Reply:
With this PR change, SparkException will not be caught and thrown directly, and nothing writing to System.err and also no exit(exitedCleanly = false) in this case,  here we need to check the thrown exception message whether it has the expected searchString or not.

Reply:
I actually wonder if throwing this exception below is actually doing anything? `thread.join()` will not throw it in the test thread, which is what would actually matter.

So it seems that to achieve what the comment above says, you should do just this instead:

```
if (!exitedCleanly) {
  message = e.getMessage
}
```

Reply:
> I actually wonder if throwing this exception below is actually doing anything?

It would give the stack trace in the test log if there is any unexpected exception. With the below, we may give exception message but no stack trace in the log during fail.
```
if (!exitedCleanly) {
  message = e.getMessage
}
```

Reply:
Then just print the exception, or fail the test explicitly with that exception somehow. At least that's explicitly doing what is expected of it.

Reply:
(e.g., by stashing the exception in a variable and doing these checks on the test thread after the `thread.join()` call.)